File: src/main/java/org/jitsi_modified/impl/neomedia/rtp/remotebitrateestimator/AimdRateControl.java
/*
 * Copyright @ 2015 Atlassian Pty Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.jitsi_modified.impl.neomedia.rtp.remotebitrateestimator;

import org.jetbrains.annotations.*;
import org.jitsi.utils.logging.DiagnosticContext;
import org.jitsi.utils.logging.TimeSeriesLogger;

/**
 * A rate control implementation based on additive increases of bitrate when no
 * over-use is detected and multiplicative decreases when over-uses are
 * detected. When we think the available bandwidth has changes or is unknown, we
 * will switch to a "slow-start mode" where we increase multiplicatively.
 *
 * webrtc/modules/remote_bitrate_estimator/aimd_rate_control.cc
 * webrtc/modules/remote_bitrate_estimator/aimd_rate_control.h
 *
 * @author Lyubomir Marinov
 */
class AimdRateControl
{
    /**
     * The <tt>TimeSeriesLogger</tt> used by the
     * <tt>RemoteBitrateEstimatorAbsSendTime</tt> class and its instances for
     * logging output.
     */
    private static final TimeSeriesLogger logger
        = TimeSeriesLogger.getTimeSeriesLogger(AimdRateControl.class);

    private static final int kDefaultRttMs = 200;

    private static final long kInitializationTimeMs = 5000;

    private static final long kFirstIncomingEstimateExpirationMs = 2 * kInitializationTimeMs;

    private static final long kLogIntervalMs = 1000;

    private static final long kMaxFeedbackIntervalMs = 1000;

    private static final long kMinFeedbackIntervalMs = 200;

    private static final int kRtcpSize = 80;

    private static final double kWithinIncomingBitrateHysteresis = 1.05;

    private final DiagnosticContext diagnosticContext;

    private float avgMaxBitrateKbps;

    private float beta;

    private boolean bitrateIsInitialized;

    /**
     * the number of time we've expired the initial incoming estimate.
     */
    private int incomingBitrateExpirations = 0;

    private long currentBitrateBps;

    private final RateControlInput currentInput
        = new RateControlInput(BandwidthUsage.kBwNormal, 0L, 0D);

    private boolean inExperiment;

    private long minConfiguredBitrateBps;

    private RateControlRegion rateControlRegion;

    private RateControlState rateControlState;

    private long rtt;

    private long timeFirstIncomingEstimate;

    private long timeLastBitrateChange;

    private long timeOfLastLog;

    private boolean updated;

    private float varMaxBitrateKbps;

    public AimdRateControl(@NotNull DiagnosticContext diagnosticContext)
    {
        reset();
        this.diagnosticContext = diagnosticContext;
    }

    private long additiveRateIncrease(
            long nowMs,
            long lastMs,
            long responseTimeMs)
    {
        if (responseTimeMs <= 0)
            throw new IllegalArgumentException("responseTimeMs");

        double beta = 0.0;

        if (lastMs > 0) {
            beta
                = Math.min((nowMs - lastMs) / (double) responseTimeMs, 1.0);
            if (inExperiment)
                beta /= 2.0;
        }

        double bitsPerFrame = (double) currentBitrateBps / 30.0;
        double packetsPerFrame = Math.ceil(bitsPerFrame / (8.0 * 1200.0));
        double avgPacketSizeBits = bitsPerFrame / packetsPerFrame;

        return (long) Math.max(1000.0, beta * avgPacketSizeBits);
    }

    private long changeBitrate(
            long currentBitrateBps,
            long incomingBitrateBps,
            long nowMs)
    {
        if (!updated)
            return this.currentBitrateBps;
        // An over-use should always trigger us to reduce the bitrate, even though
        // we have not yet established our first estimate. By acting on the over-use,
        // we will end up with a valid estimate.
        if (!bitrateIsInitialized
                && currentInput.bwState != BandwidthUsage.kBwOverusing)
        {
            return this.currentBitrateBps;
        }

        updated = false;
        changeState(currentInput, nowMs);

        // Calculated here because it's used in multiple places.
        float incomingBitrateKbps = incomingBitrateBps / 1000.0F;
        // Calculate the max bit rate std dev given the normalized variance and
        // the current incoming bit rate.
        float stdMaxBitRate
            = (float) Math.sqrt(varMaxBitrateKbps * avgMaxBitrateKbps);

        switch (rateControlState)
        {
        case kRcHold:
            break;

        case kRcIncrease:
        {
            if (avgMaxBitrateKbps >= 0F
                    && incomingBitrateKbps
                        > avgMaxBitrateKbps + 3F * stdMaxBitRate)
            {
                changeRegion(RateControlRegion.kRcMaxUnknown, nowMs);
                avgMaxBitrateKbps = -1F;
            }
            if (rateControlRegion == RateControlRegion.kRcNearMax)
            {
                // Approximate the over-use estimator delay to 100 ms.
                long responseTime = rtt + 100;
                long additiveIncreaseBps
                    = additiveRateIncrease(
                            nowMs,
                            timeLastBitrateChange,
                            responseTime);

                currentBitrateBps += additiveIncreaseBps;
            }
            else // kRcMaxUnknown || kRcAboveMax
            {
                long multiplicativeIncreaseBps
                    = multiplicativeRateIncrease(
                            nowMs,
                            timeLastBitrateChange,
                            currentBitrateBps);

                currentBitrateBps += multiplicativeIncreaseBps;
            }

            timeLastBitrateChange = nowMs;
            break;
        }
        case kRcDecrease:
        {
            bitrateIsInitialized = true;
            if (incomingBitrateBps < minConfiguredBitrateBps)
            {
                currentBitrateBps = minConfiguredBitrateBps;
            }
            else
            {
                // Set bit rate to something slightly lower than max to get rid
                // of any self-induced delay.
                currentBitrateBps = (long) (beta * incomingBitrateBps + 0.5);
                if (currentBitrateBps > this.currentBitrateBps)
                {
                    // Avoid increasing the rate when over-using.
                    if (rateControlRegion != RateControlRegion.kRcMaxUnknown)
                    {
                        currentBitrateBps
                            = (long) (beta * avgMaxBitrateKbps * 1000F + 0.5F);
                    }
                    currentBitrateBps
                        = Math.min(currentBitrateBps, this.currentBitrateBps);
                }
                changeRegion(RateControlRegion.kRcNearMax, nowMs);

                if (incomingBitrateKbps
                        < avgMaxBitrateKbps - 3F * stdMaxBitRate)
                {
                    avgMaxBitrateKbps = -1F;
                }

                updateMaxBitRateEstimate(incomingBitrateKbps);
            }
            // Stay on hold until the pipes are cleared.
            changeState(RateControlState.kRcHold, nowMs);
            timeLastBitrateChange = nowMs;
            break;
        }

        default:
            throw new IllegalStateException("rateControlState");
        }
        if ((incomingBitrateBps > 100000L || currentBitrateBps > 150000L)
                && currentBitrateBps > 1.5 * incomingBitrateBps)
        {
            // Allow changing the bit rate if we are operating at very low rates
            // Don't change the bit rate if the send side is too far off
            currentBitrateBps = this.currentBitrateBps;
            timeLastBitrateChange = nowMs;
        }
        return currentBitrateBps;
    }

    private void changeRegion(RateControlRegion region, long nowMs)
    {
        if (rateControlRegion == region)
        {
            return;
        }

        rateControlRegion = region;

        if (logger.isTraceEnabled())
        {
            logger.trace(diagnosticContext
                    .makeTimeSeriesPoint("aimd_region", nowMs)
                    .addField("aimd_id", hashCode())
                    .addField("region", region));
        }
    }

    private void changeState(RateControlInput input, long nowMs)
    {
        switch (currentInput.bwState)
        {
        case kBwNormal:
            if (rateControlState == RateControlState.kRcHold)
            {
                timeLastBitrateChange = nowMs;
                changeState(RateControlState.kRcIncrease, nowMs);
            }
            break;
        case kBwOverusing:
            if (rateControlState != RateControlState.kRcDecrease)
            {
                changeState(RateControlState.kRcDecrease, nowMs);
            }
            break;
        case kBwUnderusing:
            changeState(RateControlState.kRcHold, nowMs);
            break;
        default:
            throw new IllegalStateException("currentInput.bwState");
        }
    }

    private void changeState(RateControlState newState, long nowMs)
    {
        if (rateControlState == newState)
        {
            return;
        }

        rateControlState = newState;

        if (logger.isTraceEnabled())
        {
            logger.trace(diagnosticContext
                    .makeTimeSeriesPoint("aimd_state", nowMs)
                    .addField("aimd_id", hashCode())
                    .addField("state", rateControlState));
        }
    }

    public long getFeedBackInterval()
    {
        // Estimate how often we can send RTCP if we allocate up to 5% of
        // bandwidth to feedback.
        long interval
            = (long)
                (kRtcpSize * 8.0 * 1000.0 / (0.05 * currentBitrateBps) + 0.5);

        return
            Math.min(
                    Math.max(interval, kMinFeedbackIntervalMs),
                    kMaxFeedbackIntervalMs);
    }

    public long getLatestEstimate()
    {
        return currentBitrateBps;
    }

    /**
     * Returns <tt>true</tt> if the bitrate estimate hasn't been changed for
     * more than an RTT, or if the <tt>incomingBitrate</tt> is more than 5%
     * above the current estimate. Should be used to decide if we should reduce
     * the rate further when over-using.
     *
     * @param timeNow
     * @param incomingBitrateBps
     * @return
     */
    public boolean isTimeToReduceFurther(long timeNow, long incomingBitrateBps)
    {
        long bitrateReductionInterval = Math.max(Math.min(rtt, 200L), 10L);

        if (timeNow - timeLastBitrateChange >= bitrateReductionInterval)
            return true;
        if (isValidEstimate())
        {
          long threshold
              = (long) (kWithinIncomingBitrateHysteresis * incomingBitrateBps);
          long bitrateDifference = getLatestEstimate() - incomingBitrateBps;

          return bitrateDifference > threshold;
        }
        return false;
    }

    /**
     * @return the number of time we've expired the initial incoming estimate.
     */
    public int getIncomingEstimateExpirations()
    {
        return incomingBitrateExpirations;
    }

    /**
     * Returns <tt>true</tt> if there is a valid estimate of the incoming
     * bitrate, <tt>false</tt> otherwise.
     *
     * @return
     */
    public boolean isValidEstimate()
    {
        return bitrateIsInitialized;
    }

    private long multiplicativeRateIncrease(
            long nowMs,
            long lastMs,
            long currentBitrateBps)
    {
        double alpha = 1.08;

        if (lastMs > -1) {
            long timeSinceLastUpdateMs = Math.min(nowMs - lastMs, 1000);

            alpha = Math.pow(alpha,  timeSinceLastUpdateMs / 1000.0);
        }
        return (long) Math.max(currentBitrateBps * (alpha - 1.0), 1000.0);
    }

    public void reset()
    {
        reset(RemoteBitrateEstimatorAbsSendTime.kDefaultMinBitrateBps);
    }

    private void reset(long minBitrateBps)
    {
        minConfiguredBitrateBps = minBitrateBps;
        currentBitrateBps = /* maxConfiguredBitrateBps */ 30000000L;
        avgMaxBitrateKbps = -1F;
        varMaxBitrateKbps = 0.4F;
        rateControlState = RateControlState.kRcHold;
        rateControlRegion = RateControlRegion.kRcMaxUnknown;
        timeLastBitrateChange = -1L;
        currentInput.bwState = BandwidthUsage.kBwNormal;
        currentInput.incomingBitRate = 0L;
        currentInput.noiseVar = 1D;
        updated = false;
        timeFirstIncomingEstimate = -1L;
        bitrateIsInitialized = false;
        beta = 0.85F;
        rtt = kDefaultRttMs;
        timeOfLastLog = -1L;
        inExperiment = false;
    }

    public void setEstimate(long bitrateBps, long nowMs)
    {
        updated = true;
        bitrateIsInitialized = true;
        currentBitrateBps = changeBitrate(bitrateBps, bitrateBps, nowMs);
    }

    public void setMinBitrate(long minBitrateBps)
    {
        minConfiguredBitrateBps = minBitrateBps;
        currentBitrateBps = Math.max(minBitrateBps, currentBitrateBps);
    }

    public void setRtt(long nowMs, long rtt)
    {
        if (logger.isTraceEnabled())
        {
            logger.trace(diagnosticContext
                    .makeTimeSeriesPoint("aimd_rtt", nowMs)
                    .addField("aimd_id", hashCode())
                    .addField("rtt", rtt));
        }

        this.rtt = rtt;
    }

    public void update(RateControlInput input, long nowMs)
    {
        if (input == null)
            throw new NullPointerException("input");

        // Set the initial bit rate value to what we're receiving the first half
        // second.
        if (!bitrateIsInitialized)
        {
            if (timeFirstIncomingEstimate < 0L)
            {
                if (input.incomingBitRate > 0L)
                    timeFirstIncomingEstimate = nowMs;
            }
            else
            {
                long timeSinceFirstIncomingEstimate = nowMs - timeFirstIncomingEstimate;
                if (timeSinceFirstIncomingEstimate > kFirstIncomingEstimateExpirationMs)
                {
                    if (input.incomingBitRate > 0L)
                    {
                        timeFirstIncomingEstimate = nowMs;
                    }
                    else
                    {
                        timeFirstIncomingEstimate = -1L;
                    }

                    incomingBitrateExpirations++;
                }
                else if (timeSinceFirstIncomingEstimate > kInitializationTimeMs
                    && input.incomingBitRate > 0L)
                {
                    currentBitrateBps = input.incomingBitRate;
                    bitrateIsInitialized = true;
                }
            }
        }

        if (updated && currentInput.bwState == BandwidthUsage.kBwOverusing)
        {
            // Only update delay factor and incoming bit rate. We always want to
            // react on an over-use.
            currentInput.noiseVar = input.noiseVar;
            currentInput.incomingBitRate = input.incomingBitRate;
        }
        else
        {
            updated = true;
            currentInput.copy(input);
        }
    }

    public long updateBandwidthEstimate(long nowMs)
    {
        currentBitrateBps
            = changeBitrate(
                    currentBitrateBps,
                    currentInput.incomingBitRate,
                    nowMs);

        if (logger.isTraceEnabled() && isValidEstimate())
        {
            logger.trace(diagnosticContext
                .makeTimeSeriesPoint("aimd_estimate", nowMs)
                .addField("aimd_id", hashCode())
                .addField("estimate_bps", currentBitrateBps)
                .addField("incoming_bps", currentInput.incomingBitRate));
        }

        if (nowMs - timeOfLastLog > kLogIntervalMs)
            timeOfLastLog = nowMs;
        return currentBitrateBps;
    }

    private void updateMaxBitRateEstimate(float incomingBitrateKbps)
    {
        float alpha = 0.05F;

        if (avgMaxBitrateKbps == -1F)
        {
            avgMaxBitrateKbps = incomingBitrateKbps;
        }
        else
        {
            avgMaxBitrateKbps
                = (1 - alpha) * avgMaxBitrateKbps + alpha * incomingBitrateKbps;
        }

        // Estimate the max bit rate variance and normalize the variance with
        // the average max bit rate.
        float norm = Math.max(avgMaxBitrateKbps, 1F);

        varMaxBitrateKbps
            = (1 - alpha) * varMaxBitrateKbps
                + alpha
                    * (avgMaxBitrateKbps - incomingBitrateKbps)
                    * (avgMaxBitrateKbps - incomingBitrateKbps)
                    / norm;
        // 0.4 ~= 14 kbit/s at 500 kbit/s
        if (varMaxBitrateKbps < 0.4F)
            varMaxBitrateKbps = 0.4F;
        // 2.5f ~= 35 kbit/s at 500 kbit/s
        if (varMaxBitrateKbps > 2.5f)
            varMaxBitrateKbps = 2.5f;
    }
}

File: jitsi-media-transform/src/main/kotlin/org/jitsi/nlj/rtp/bandwidthestimation2/AimdRateControl.kt
/*
 * Copyright @ 2019 - present 8x8, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
// This file uses WebRTC's naming style for enums
@file:Suppress("ktlint:standard:enum-entry-name-case", "ktlint:standard:property-naming")

package org.jitsi.nlj.rtp.bandwidthestimation2

import edu.umd.cs.findbugs.annotations.SuppressFBWarnings
import org.jitsi.nlj.util.Bandwidth
import org.jitsi.nlj.util.bps
import org.jitsi.nlj.util.bytes
import org.jitsi.nlj.util.div
import org.jitsi.nlj.util.kbps
import org.jitsi.nlj.util.max
import org.jitsi.nlj.util.per
import org.jitsi.nlj.util.times
import org.jitsi.utils.NEVER
import org.jitsi.utils.coerceIn
import org.jitsi.utils.div
import org.jitsi.utils.durationOfDoubleSeconds
import org.jitsi.utils.max
import org.jitsi.utils.ms
import org.jitsi.utils.secs
import org.jitsi.utils.times
import org.jitsi.utils.toDouble
import java.time.Duration
import java.time.Instant
import kotlin.math.ceil
import kotlin.math.min
import kotlin.math.pow

/**
 * A rate control implementation based on additive increases of bitrate when no
 * over-use is detected and multiplicative decreases when over-uses are
 * detected. When we think the available bandwidth has changes or is unknown, we
 * will switch to a "slow-start mode" where we increase multiplicatively.
 *
 * Based on WebRTC modules/remote_bitrate_estimator/aimd_rate_control.{h,cc} in
 * WebRTC tag branch-heads/7204 (Chromium 138).
 */
class AimdRateControl(private val sendSide: Boolean = false) {

    private var minConfiguredBitrate: Bandwidth = kCongestionControllerMinBitrate
    private var maxConfiguredBitrate: Bandwidth = 30000.kbps
    private var currentBitrate: Bandwidth = maxConfiguredBitrate
    private var latestEstimatedThroughput: Bandwidth = currentBitrate

    val linkCapacity = LinkCapacityEstimator()

    // Omitted:
    // val networkEstimate: NetworkStateEstimate? = null
    // As far as I can tell it is not ever set in current Chromium?

    var rateControlState: RateControlState = RateControlState.kRcHold
        private set

    private var timeLastBitrateChange: Instant = NEVER

    @SuppressFBWarnings("URF_UNREAD_FIELD")
    private var timeLastBitrateDecrease: Instant = NEVER
    private var timeFirstThroughputEstimate: Instant = NEVER

    private var bitrateIsInitialized: Boolean = false
    private var beta: Double = kDefaultBackoffFactor
    var inAlr: Boolean = false
    var rtt: Duration = kDefaultRtt

    // TODO: field trials: remove code that checks these

    // Allow the delay based estimate to only increase as long as application
    // limited region (alr) is not detected.
    private val noBitrateIncreaseInAlr: Boolean = false

    // Only tested when networkEstimate != null
    // private val disableEstimateBoundedIncrease: Boolean = false
    // private val useCurrentEstimateAsMinUpperBound = true

    private var lastDecrease: Bandwidth? = null

    fun setStartBitrate(startBitrate: Bandwidth) {
        currentBitrate = startBitrate
        latestEstimatedThroughput = startBitrate
        bitrateIsInitialized = true
    }

    fun setMinBitrate(minBitrate: Bandwidth) {
        minConfiguredBitrate = minBitrate
        currentBitrate = max(minBitrate, currentBitrate)
    }

    fun validEstimate(): Boolean = bitrateIsInitialized

    fun getFeedbackInterval(): Duration {
        // Estimate how often we can send RTCP if we allocate up to 5% of bandwidth
        // to feedback.
        val kRtcpSize = 80.bytes
        val rtcpBitrate = currentBitrate * 0.05
        val interval = kRtcpSize / rtcpBitrate
        val kMinFeedbackInterval = 200.ms
        val kMaxFeedbackInterval = 1000.ms
        return interval.coerceIn(kMinFeedbackInterval, kMaxFeedbackInterval)
    }

    fun timeToReduceFurther(atTime: Instant, estimatedThroughput: Bandwidth): Boolean {
        val bitrateReductionInterval = rtt.coerceIn(10.ms, 200.ms)
        if (Duration.between(timeLastBitrateChange, atTime) >= bitrateReductionInterval) {
            return true
        }
        if (validEstimate()) {
            // TODO(terelius/holmer): Investigate consequences of increasing
            // the threshold to 0.95 * LatestEstimate().
            val threshold = 0.5 * latestEstimate()
            return estimatedThroughput < threshold
        }
        return false
    }

    fun initialTimeToReduceFurther(atTime: Instant): Boolean {
        return validEstimate() &&
            timeToReduceFurther(atTime, latestEstimate() / 2 - 1.bps)
    }

    fun latestEstimate() = currentBitrate

    fun update(input: RateControlInput, atTime: Instant): Bandwidth {
        // Set the initial bit rate value to what we're receiving the first half
        // second.
        // TODO(bugs.webrtc.org/9379): The comment above doesn't match to the code.
        if (!bitrateIsInitialized) {
            val kInitializionTime = 5.secs
            check(kBitrateWindow <= kInitializionTime)
            if (timeFirstThroughputEstimate == NEVER) {
                if (input.estimatedThroughput != null) {
                    timeFirstThroughputEstimate = atTime
                }
            } else if (Duration.between(timeFirstThroughputEstimate, atTime) > kInitializionTime) {
                input.estimatedThroughput?.let {
                    currentBitrate = it
                    bitrateIsInitialized = true
                }
            }
        }
        changeBitrate(input, atTime)
        return currentBitrate
    }

    fun setEstimate(bitrate: Bandwidth, atTime: Instant) {
        bitrateIsInitialized = true
        val prevBitrate = currentBitrate
        currentBitrate = clampBitrate(bitrate)
        timeLastBitrateChange = atTime
        if (currentBitrate < prevBitrate) {
            timeLastBitrateDecrease = atTime
        }
    }

    fun getNearMaxIncreaseRateBpsPerSecond(): Double {
        check(currentBitrate != 0.bps)
        val kFrameInterval = 1.secs / 30.0
        val frameSize = (currentBitrate * kFrameInterval).toWholeBytes()
        val kPacketSize = 1200.bytes
        val packetsPerFrame = ceil(frameSize / kPacketSize)
        val avgPacketSize = (frameSize / packetsPerFrame).toWholeBytes()

        // Approximate the over-use estimator delay to 100 ms.
        var responseTime = rtt + 100.ms

        responseTime = responseTime * 2
        val increaseRateBpsPerSecond = (avgPacketSize.per(responseTime)).bps.toDouble()
        val kMinIncreaseRateBpsPerSecond = 4000.0
        return increaseRateBpsPerSecond.coerceAtLeast(kMinIncreaseRateBpsPerSecond)
    }

    fun getExpectedBandwidthPeriod(): Duration {
        val kMinPeriod = 2.secs
        val kDefaultPeriod = 3.secs
        val kMaxPeriod = 50.secs

        val increaseRateBpsPerSecond = getNearMaxIncreaseRateBpsPerSecond()
        if (lastDecrease == null) {
            return kDefaultPeriod
        }

        val timeToRecoverDecreaseSeconds =
            lastDecrease!!.bps / increaseRateBpsPerSecond

        val period = durationOfDoubleSeconds(timeToRecoverDecreaseSeconds)
        return period.coerceIn(kMinPeriod, kMaxPeriod)
    }

    private fun changeBitrate(input: RateControlInput, atTime: Instant) {
        var newBitrate: Bandwidth? = null
        val estimatedThroughput = input.estimatedThroughput ?: latestEstimatedThroughput
        input.estimatedThroughput?.let { latestEstimatedThroughput = it }

        // An over-use should always trigger us to reduce the bitrate, even though
        // we have not yet established our first estimate. By acting on the over-use,
        // we will end up with a valid estimate.
        if (!bitrateIsInitialized && input.bwState != BandwidthUsage.kBwOverusing) {
            return
        }

        changeState(input, atTime)

        when (rateControlState) {
            RateControlState.kRcHold -> Unit

            RateControlState.kRcIncrease -> {
                if (estimatedThroughput > linkCapacity.upperBound()) {
                    linkCapacity.reset()
                }

                // We limit the new bitrate based on the troughput to avoid unlimited
                // bitrate increases. We allow a bit more lag at very low rates to not too
                // easily get stuck if the encoder produces uneven outputs.

                var increaseLimit = 1.5 * estimatedThroughput + 10.kbps
                if (sendSide && inAlr && noBitrateIncreaseInAlr) {
                    // TODO this is dead code because we're not using the noBitrateIncreaseInAlr field trial
                    // Do not increase the delay based estimate in alr since the estimator
                    // will not be able to get transport feedback necessary to detect if
                    // the new estimate is correct.
                    // If we have previously increased above the limit (for instance due to
                    // probing), we don't allow further changes.
                    increaseLimit = currentBitrate
                }

                if (currentBitrate < increaseLimit) {
                    var increasedBitrate = Bandwidth.MINUS_INFINITY
                    if (linkCapacity.hasEstimate()) {
                        // The link_capacity estimate is reset if the measured throughput
                        // is too far from the estimate. We can therefore assume that our
                        // target rate is reasonably close to link capacity and use additive
                        // increase.
                        val additiveIncrease =
                            additiveRateIncrease(atTime, timeLastBitrateChange)
                        increasedBitrate = currentBitrate + additiveIncrease
                    } else {
                        // If we don't have an estimate of the link capacity, use faster ramp
                        // up to discover the capacity.
                        val multiplicativeIncrease = multiplicativeRateIncrease(
                            atTime,
                            timeLastBitrateChange,
                            currentBitrate
                        )
                        increasedBitrate = currentBitrate + multiplicativeIncrease
                    }
                    newBitrate = increasedBitrate.coerceAtMost(increaseLimit)
                }
                timeLastBitrateChange = atTime
            }
            RateControlState.kRcDecrease -> {
                var decreasedBitrate = Bandwidth.INFINITY

                // Set bit rate to something slightly lower than the measured throughput
                // to get rid of any self-induced delay.
                decreasedBitrate = estimatedThroughput * beta
                if (decreasedBitrate > 5.kbps) {
                    decreasedBitrate -= 5.kbps
                }
                if (decreasedBitrate > currentBitrate) {
                    // TODO(terelius): The link_capacity estimate may be based on old
                    // throughput measurements. Relying on them may lead to unnecessary
                    // BWE drops.
                    if (linkCapacity.hasEstimate()) {
                        decreasedBitrate = linkCapacity.estimate!! * beta
                    }
                }
                // Avoid increasing the rate when over-using.
                if (decreasedBitrate < currentBitrate) {
                    newBitrate = decreasedBitrate
                }

                if (bitrateIsInitialized && estimatedThroughput < currentBitrate) {
                    if (newBitrate == null) {
                        lastDecrease = Bandwidth.ZERO
                    } else {
                        lastDecrease = currentBitrate - newBitrate
                    }
                }
                if (estimatedThroughput < linkCapacity.lowerBound()) {
                    // The current throughput is far from the estimated link capacity. Clear
                    // the estimate to allow an immediate update in onOveruseDetected.
                    linkCapacity.reset()
                }

                bitrateIsInitialized = true
                linkCapacity.onOveruseDetected(estimatedThroughput)
                // Stay on hold until the pipes are cleared.
                rateControlState = RateControlState.kRcHold
                timeLastBitrateChange = atTime
                timeLastBitrateDecrease = atTime
            }
        }
        currentBitrate = clampBitrate(newBitrate ?: currentBitrate)
    }

    private fun clampBitrate(newBitrate: Bandwidth): Bandwidth {
        /* Skipping some conditions related to networkEstimate != null */

        return newBitrate.coerceAtLeast(minConfiguredBitrate)
    }

    private fun multiplicativeRateIncrease(atTime: Instant, lastTime: Instant, currentBitrate: Bandwidth): Bandwidth {
        var alpha = 1.08
        if (lastTime != NEVER) {
            val timeSinceLastUpdate = Duration.between(lastTime, atTime)
            alpha = alpha.pow(min(timeSinceLastUpdate.toDouble(), 1.0))
        }
        val multiplicativeIncrease =
            max(currentBitrate * (alpha - 1.0), 1000.bps)
        return multiplicativeIncrease
    }

    private fun additiveRateIncrease(atTime: Instant, lastTime: Instant): Bandwidth {
        val timePeriodSeconds = Duration.between(lastTime, atTime).toDouble()
        val dataRateIncreaseBps = getNearMaxIncreaseRateBpsPerSecond() * timePeriodSeconds
        return dataRateIncreaseBps.bps
    }

    private fun changeState(input: RateControlInput, atTime: Instant) {
        when (input.bwState) {
            BandwidthUsage.kBwNormal ->
                if (rateControlState == RateControlState.kRcHold) {
                    timeLastBitrateChange = atTime
                    rateControlState = RateControlState.kRcIncrease
                }
            BandwidthUsage.kBwOverusing ->
                if (rateControlState != RateControlState.kRcDecrease) {
                    rateControlState = RateControlState.kRcDecrease
                }
            BandwidthUsage.kBwUnderusing ->
                rateControlState = RateControlState.kRcHold
        }
    }

    enum class RateControlState {
        kRcHold,
        kRcIncrease,
        kRcDecrease
    }

    companion object {
        private val kDefaultRtt = 200.ms
        private const val kDefaultBackoffFactor = 0.85
    }
}
